l3kernel/l3bitset.dtx

```diff
@@ -0,0 +1,551 @@
+%    }
+%  \end{verbatim}
+%  it is possible to set bit 2 by using
+%  |\bitset_set_key:Nn \l_pdfannot_F_bitset {Print}|
+%  If a key is unknown a warning is issued.
+%  It is not needed to give  all index positions a name.
+% \begin{function}[added = 2020-12-13]
+%   {\bitset_set_key:Nn, \bitset_set_key:cn,
+%    \bitset_gset_key:Nn,\bitset_gset_key:cn }
+%   \begin{syntax}
+%     \cs{bitset_set_key:nn}   \meta{bitset var}  \Arg{key}\\
+%     \cs{bitset_gset_key:nn}  \meta{bitset var}  \Arg{key}
+%   \end{syntax}
+% This sets the bit of the position described by key to \texttt{1}.
+%
+% If needed the length of the bit vector is enlarged.
+% \end{function}
+% \begin{function}[added = 2020-12-13]
+%   {\bitset_clear_key:Nn, \bitset_clear_key:cn,
+%    \bitset_gclear_key:Nn, \bitset_clear_key:cn,}
+%   \begin{syntax}
+%     \cs{bitset_clear_key:Nn}   \meta{bitset var}  \Arg{key}\\
+%     \cs{bitset_gclear_key:Nn}  \meta{bitset var}  \Arg{key}
+%   \end{syntax}
+% This sets the bit of the position described by key to 0.
+% If the leading (left most) bit is cleared, zeros are not
+% trimmed but stay in the bit vector.
+% \end{function}
+% \begin{function}[EXP,added = 2020-12-14]
+%   {\bitset_key:Nn}
+%   \begin{syntax}
+%     \cs{bitset_key:Nn}     \meta{bitset var}  \Arg{key}
+%   \end{syntax}
+% This outputs \texttt{1} if the bit is set, \texttt{0} otherwise.
+% \end{function}
+% \end{documentation}
+%
+% \begin{implementation}
+% \section{\pkg{l3bitset} implementation}
+% \TestFiles{m3bitset001,m3bitset002}
+%    \begin{macrocode}
+%<*package>
+%    \end{macrocode}
+%
+%    \begin{macrocode}
+%<@@=bitset>
+%    \end{macrocode}
+% A bitset is a string variable.
+%  \begin{macro}
+%    {
+%      \bitset_new:N,  \bitset_new:c
+%    }
+%    \begin{macrocode}
+\cs_new_protected:Npn \bitset_new:N #1
+  {
+    \str_new:N #1
+  }
+\cs_generate_variant:Nn \bitset_new:N {c}
+%    \end{macrocode}
+% \end{macro}
+% \begin{macro}[pTF]
+%   {
+%     \bitset_if_exist:N, \bitset_if_exist:c
+%   }
+%
+% Existence tests.
+%    \begin{macrocode}
+\prg_new_eq_conditional:NNn
+  \bitset_if_exist:N \str_if_exist:N { p , T , F , TF }
+\prg_new_eq_conditional:NNn
+  \bitset_if_exist:c \str_if_exist:c
+  { p , T , F , TF }
+%    \end{macrocode}
+% \end{macro}
+
+% \begin{macro}
+%   {
+%     \bitset_set:Nn, \bitset_set:cn,
+%     \bitset_gset:Nn, \bitset_gset:cn,
+%   }
+% A bit is set by either extending the string or by splitting it and
+% then inserting an 1. It is not checked if the value was already 1.
+%    \begin{macrocode}
+% #1 name, #2 index (integer expression,zero based)
+\cs_new_protected:Npn \bitset_set:Nn #1 #2
+  {
+    \int_compare:nNnT {#2 } > { -1 }
+      {
+        \int_compare:nNnTF {\str_count:N #1 } < { #2 + 1 }
+          {
+            %extend the str
+            \int_step_inline:nn { #2 - \str_count:N #1 }
+              {
+                \str_put_left:Nn #1 { 0 }
+              }
+            \str_put_left:Nn #1 { 1 }
```

Comment:
I haven't measured this, but something like

```
\str_put_left:Nn {
  1
  \prg_replicate:nn { #2 - \str_count:N #1 }
}
```
feels like it should be faster and is IMO more readable.

Reply:
Probably  `\str_put_left:Ne` or Nx? but beside this you are quite right, it looks more readable.